java/src/test/java/org/rocksdb/SequenceNumberAllocationTest.java

```diff
@@ -0,0 +1,198 @@
+
+    @Test
+    public void testSeqnoFlushAndClose(WriteOptions writeOptions)
+            throws RocksDBException
+    {
+        int count;
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            count = fillDatabase(db, writeOptions, 0, 15);
+            db.flush(new FlushOptions());
+            db.closeE();
+        }
+
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            assertTrue("Database should not be empty", db.getLatestSequenceNumber() != 0);
+            assertEquals("Sequence number changed between closing and reopening the database", count, db.getLatestSequenceNumber());
+        }
+    }
+
+    @Test
+    public void testSeqnoCloseWithoutFlush(WriteOptions writeOptions)
+            throws RocksDBException
+    {
+        int count;
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            count = fillDatabase(db, writeOptions, 0, 30);
+            db.closeE();
+        }
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            assertTrue("Database should not be empty", db.getLatestSequenceNumber() != 0);
+            assertEquals("Sequence number changed between closing and reopening the database", count, db.getLatestSequenceNumber());
+        }
+    }
+
+    @Test
+    public void testSeqnoThreadInterruptAfterFlush(WriteOptions writeOptions)
+            throws InterruptedException
+    {
+        ExecutorService executor = Executors.newFixedThreadPool(1);
+        AtomicLong sequenceNumberAfterFlush = new AtomicLong(0);
+        AtomicBoolean cancellable = new AtomicBoolean(false);
+        Options options = new Options().setCreateIfMissing(true);
+
+        Future<?> task = executor.submit(() -> {
+            try {
+                RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath());
+                int count = fillDatabase(db, writeOptions, 0, 10);
+                assertEquals("Expected 10 values in the database",10, count);
+                db.flush(new FlushOptions());
+                sequenceNumberAfterFlush.set(db.getLatestSequenceNumber());
+                assertEquals("Sequence number after flush should be 10", 10, sequenceNumberAfterFlush.get());
+                cancellable.set(true);
+                while(true) {
+                    // Wait indefinitely
+                }
+            }
+            catch (RocksDBException e) {
+                fail("Encountered exception:" + e);
+            }
+        });
+
+        // Enough time to finish performing database operations
+        sleep(5 * 1000);
```

Comment:
This seems a little unpredictable, is there a better way of knowing when the database operations have completed?

Reply:
Thanks! Updated with CountdownLatches!